src/drivers/ins/microstrain/MicroStrain.cpp

```diff
@@ -0,0 +1,623 @@
+				break;
+			}
+		}
+
+		if (unique) {_supported_descriptor_sets[j] = descriptor_set; _supported_desc_set_len++;}
+	}
+
+	return res;
+}
+
+bool MicroStrain::supports_descriptor(uint8_t descriptor_set, uint8_t field_descriptor)
+{
+	bool check_flag = false;
+
+	for (uint16_t i = 0; i < _supported_desc_set_len; i++) {
+		if (_supported_descriptor_sets[i] == descriptor_set) {
+			check_flag = true;
+			break;
+		}
+	}
+
+	if (!check_flag) {return check_flag;}
+
+	const uint16_t full_descriptor = (descriptor_set << 8) | field_descriptor;
+
+	check_flag = false;
+
+	for (uint16_t i = 0; i < _supported_desc_len; i++) {
+		if (_supported_descriptors[i] == full_descriptor) {
+			check_flag = true;
+			break;
+		}
+	}
+
+	return check_flag;
+}
+
+mip_cmd_result MicroStrain::get_base_rate(uint8_t descriptor_set, uint16_t *base_rate)
+{
+
+
+	if (supports_descriptor(MIP_3DM_CMD_DESC_SET, MIP_CMD_DESC_3DM_GET_BASE_RATE)) {
+		return mip_3dm_get_base_rate(&device, descriptor_set, base_rate);
+
+	} else {
+		switch (descriptor_set) {
+		case MIP_SENSOR_DATA_DESC_SET:
+			return mip_3dm_imu_get_base_rate(&device, base_rate);
+
+		default:
+			return MIP_NACK_INVALID_PARAM;
+		}
+	}
+
+}
+
+mip_cmd_result MicroStrain::configure_imu_message_format()
+{
+	uint8_t num_imu_descriptors = 0;
+	mip_descriptor_rate imu_descriptors[4];
+
+	uint16_t base_rate;
+	// Get the base rate
+	mip_cmd_result res = get_base_rate(MIP_SENSOR_DATA_DESC_SET, &base_rate);
+
+	PX4_INFO("The sensor base rate is %d", base_rate);
+
+	if (res != MIP_ACK_OK) {
+		PX4_ERR("ERROR: Could not get the base rate");
+		return res;
+	}
+
+	uint16_t imu_decimation = base_rate / (uint16_t)_param_ms_imu_rate_hz.get();
+	uint16_t mag_decimation = base_rate / (uint16_t)_param_ms_mag_rate_hz.get();
+	uint16_t baro_decimation = base_rate / (uint16_t)_param_ms_baro_rate_hz.get();
+
+	if (supports_descriptor(MIP_SENSOR_DATA_DESC_SET, MIP_DATA_DESC_SENSOR_ACCEL_SCALED)
+	    && _param_ms_imu_rate_hz.get() > 0) {
+		imu_descriptors[num_imu_descriptors++] = mip_descriptor_rate { MIP_DATA_DESC_SENSOR_ACCEL_SCALED, imu_decimation};
+	}
+
+	if (supports_descriptor(MIP_SENSOR_DATA_DESC_SET, MIP_DATA_DESC_SENSOR_GYRO_SCALED)
+	    && _param_ms_imu_rate_hz.get() > 0) {
+		imu_descriptors[num_imu_descriptors++] = mip_descriptor_rate { MIP_DATA_DESC_SENSOR_GYRO_SCALED, imu_decimation};
+	}
+
+	if (supports_descriptor(MIP_SENSOR_DATA_DESC_SET, MIP_DATA_DESC_SENSOR_MAG_SCALED)
+	    && _param_ms_mag_rate_hz.get() > 0) {
+		imu_descriptors[num_imu_descriptors++] = mip_descriptor_rate { MIP_DATA_DESC_SENSOR_MAG_SCALED, mag_decimation};
+	}
+
+	if (supports_descriptor(MIP_SENSOR_DATA_DESC_SET, MIP_DATA_DESC_SENSOR_PRESSURE_SCALED)
+	    && _param_ms_baro_rate_hz.get() > 0) {
+		imu_descriptors[num_imu_descriptors++] = mip_descriptor_rate { MIP_DATA_DESC_SENSOR_PRESSURE_SCALED, baro_decimation};
+	}
+
+
+	// // Write the settings
+	res = mip_3dm_write_message_format(&device, MIP_SENSOR_DATA_DESC_SET, num_imu_descriptors,
```

Comment:
This is another one with multiple different versions of the function for different devices. See [this function](https://github.com/LORD-MicroStrain/microstrain_inertial_driver_common/blob/main/src/utils/mip/ros_mip_device_main.cpp#L266) in the ROS driver